src/app/modules/form/admin-form/admin-form.utils.ts

```diff
@@ -115,14 +114,67 @@ export const assertHasReadPermissions = (
       (allowedUser) => allowedUser.email === user.email,
     )
 
-    if (!hasReadPermissions) {
-      return err(
-        new ForbiddenFormError(
-          `User ${user.email} not authorized to perform read operation on Form ${form._id} with title: ${form.title}.`,
-        ),
-      )
+    return hasReadPermissions
+      ? ok(true)
+      : err(
+          new ForbiddenFormError(
+            `User ${user.email} not authorized to perform read operation on Form ${form._id} with title: ${form.title}.`,
+          ),
+        )
+  })
+}
+
+/**
+ * Asserts that the given user has delete permissions for the form.
+ * @returns ok(true) if given user has delete permissions
+ * @returns err(FormDeletedError) if form has already been archived
+ * @returns err(ForbiddenFormError) if user does not have delete permissions
+ */
+export const assertHasDeletePermissions = (
+  user: IUserSchema,
+  form: IPopulatedForm,
+): FormPermissionResult => {
+  return isFormActive(form).andThen(() => {
```

Comment:
personally I would rather give each of these functions a single responsibility, rather than having all of them individually check for `isFormActive`. perhaps the controller function can look like 
```
return isFormActive(req.form)
  .andThen(() => {
    switch (requiredPermission) { ... }
  })
```

Reply:
Makes sense, good catch! Pushed new commits that does this. Please rereview